### ISSUE TITLE
Unmarshal requires read?

### ISSUE THREAD

**Issue by jdbruijn** [spf13/viper#882]
I'm using viper to unmarshal into my structure. I now have the following but it looks like I explicitly need to call `ReadInConfig` before unmarshal returns any values. I didn't find this explicitly in the docs as the examples just call `Unmarshal`, so if this is indeed required I can create a PR to fix the docs perhaps.

```go
package config

import (
	"fmt"

	"github.com/spf13/viper"
)

// Config configuration schema.
type Config struct {
	Server struct {
		// Port to listen on.
		ListenPort uint16 `mapstructure:"listen-port"`
		// Path of a file containing the certificate (chain) of the server in PEM
		// format.
		Certificate string `mapstructure:"certificate"`
		// Path of a file with the secret key in PEM format.
		CertificateKey string `mapstructure:"certificate-key"`
	} `mapstructure:"server"`

	HTTPService struct {
		Address string
		Port    uint16
	} `mapstructure:"http-service"`
}

// Read reads a configuration file.
func Read(configFile string) {
	viper.SetConfigFile(configFile)

	var config Config
	// err := viper.ReadInConfig()
	err := viper.Unmarshal(&config)
	if err != nil {
		panic(fmt.Errorf("unable to decode into struct, %v", err))
	}

	fmt.Println("Using config from", viper.ConfigFileUsed())
	fmt.Printf("config %+v\n", config)
	viper.Debug()
}

```

- Output without `ReadInConfig`
   ```
   config {Server:{ListenPort:0 Certificate: CertificateKey:} HTTPService:{Address: Port:0}}
   ```
- Output with `ReadInConfig`
  ```
  config {Server:{ListenPort:5684 Certificate:./test/certificates/servers/1-ca-chain.pem CertificateKey:./test/certificates/servers/1-ca.pem} HTTPService:{Address:localhost Port:8080}}
   ```

**Comment by sagikazarmark** [comment]
The documentation explicitly shows how configuration file reading works:

https://github.com/spf13/viper#reading-config-files

Unmarshaling doesn't require reading a config file per se, it works with flags, defaults and environment variables as well.

**Comment by jdbruijn** [comment]
Thanks for your quick reply!

The only configuration I'm using is from a JSON-file, so no env vars or defaults. And below is most of the setup I used when I ran into this. I did read the docs and I am using the unmarshal example (https://github.com/spf13/viper#unmarshaling) as basis, which doesn't have the `ReadInConfig`.

Let me rephrase the question: Do I need to explicitly call `ReadInConfig` when I want to use a configuration from that file?


My code.

```go
package config

import (
	"fmt"

	"github.com/spf13/viper"
	"gopkg.in/dealancer/validate.v2"
)

// Config configuration schema.
type Config struct {
	Server struct {
		// Port to listen on.
		ListenPort uint16 `mapstructure:"listen-port" validate:"gte=1"`
		// Path of a file containing the certificate (chain) of the server in PEM
		// format.
		Certificate string `mapstructure:"certificate" validate:"format=file"`
		// Path of a file with the secret key in PEM format.
		CertificateKey string `mapstructure:"certificate-key" validate:"format=file"`
	} `mapstructure:"server"`

	HTTPService struct {
		Address string `mapstructure:"address" validate:"format=hostname_rfc1123"`
		Port    uint16 `mapstructure:"port" validate:"gte=1"`
	} `mapstructure:"http-service"`
}

func (c *Config) validate() {
	if err := validate.Validate(&c); err != nil {
		panic(err)
	}
}

// Read reads a configuration file.
func Read(configFile string) {
	viper.SetConfigFile(configFile)

	var config Config
	err := viper.ReadInConfig()
	err = viper.Unmarshal(&config)
	if err != nil {
		panic(fmt.Errorf("unable to decode into struct, %v", err))
	}

	fmt.Println("Using config from", viper.ConfigFileUsed())
	fmt.Printf("config %+v\n", config)
	config.validate()
}
```

My configuration file.

```json
{
  "server": {
    "listen-port": 5684,
    "certificate": "./test/certificates/servers/1-ca-chain.pem",
    "certificate-key": "./test/certificates/servers/1-ca.pem"
  },
  "http-service": {
    "address": "localhost",
    "port": 8080
  }
}
```

**Comment by sagikazarmark** [comment]
> The only configuration I'm using is from a JSON-file, so no env vars or defaults

The point is that Viper can read configuration from several sources, file is only one of them. Unmarshal is a separate concept, assuming that you are familiar with Viper sources, thus the documentation examples only show how Unmarshal works, not how the different config sources do.

> Let me try to rephrase the question: Do I need to explicitly call `ReadInConfig` when I want to use a configuration from that file?

Yes, you do.

**Comment by jdbruijn** [comment]
I get it now, thanks! I appreciate the help!
